### PR TITLE
review

### PR DESCRIPTION
**Pull request by yershovkiril**

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    redirect_to cart_path
+  end
+  
+  def update
+    if @order.update(order_params)
+      if params.key?(:checkout)
+        redirect_to checkout_index_path
+      else
+        redirect_to cart_path
+      end
```
Comment (yamnich) [inline app/controllers/orders_controller.rb]: move to method

Comment (yamnich) [inline app/controllers/orders_controller.rb]: 1. move to method smth like redirect_path

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  
+  def update_password
+    if @user.update_with_password(user_params)
+      sign_in @user, bypass: true
+      flash[:notice] = t('user.flashes.password_updated')
+      redirect_to edit_user_path
+    else
+      render :edit
+    end
+  end
```
Comment (yamnich) [inline app/controllers/users_controller.rb]: probably we can use devise controller here

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
+class UsersController < ApplicationController
+  before_action :authenticate_user!
+  before_action :find_user, except: [:show]
+  
+  def show
+    @history = OrderHistory.new(current_user)
```
Comment (yamnich) [inline app/controllers/users_controller.rb]: it is better to place it into orders controller

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      redirect_to edit_user_path
+    else
+      render :edit
+    end
+  end
+  
+  private
+  
+  def find_user
+    @user = current_user
```
Comment (yamnich) [inline app/controllers/users_controller.rb]: why can't we use current_user ?

File: app/forms/checkout_form.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    case step
+    when :address then bind_address
+    when :delivery then bind_delivery
+    when :payment then bind_payment
+    end
+  end
+
+  def save_by_step
+    case step
+    when :address then shipping_address.save && billing_address.save
```
Comment (yamnich) [inline app/forms/checkout_form.rb]: shipping_address.save && billing_address.save move to separate method

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+module ApplicationHelper
+  def full_title(page_title	=	'')
```
Comment (yamnich) [inline app/helpers/application_helper.rb]: whitespaces

File: app/models/ability.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    if user && user.admin
+      can :access, :rails_admin
+      can :manage, :all
+    else
+      can :read, [Book, Category, Author, Review]
+      can :manage, Order
+      can :manage, OrderBook
+      can :manage, User
+      can :create, Review
+      can :update, Review
```
Comment (yamnich) [inline app/models/ability.rb:15]: should be added rules for logged in and not logged in user

File: app/models/author.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class Author < ActiveRecord::Base
+  validates :first_name, :last_name, presence: true
+  has_many :books
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
```
Comment (yamnich) [inline app/models/author.rb]: this is better place into decorator

File: app/views/books/show.html.haml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  .row
+    .col-xs-4
+      = image_tag(@book.image, class: 'img-responsive preview')
+    .col-xs-8  
+      .caption-full
+        %h1.title= @book.title
+        %span.author by #{@book.author_name}
+        %p= @book.full_description
+        %p.price-p
+          %span.price Price:
```
Comment (yamnich) [inline app/views/books/show.html.haml]: I18n

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<p>Welcome <%= @email %>!</p>
```
Comment (yamnich) [inline app/views/devise/mailer/confirmation_instructions.html.erb]: remove extra devise views

File: app/views/orders/_order_details.html.haml [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+      %th= t('order.number')
+      %th= t('order.completed_at')
+      %th= t('order.total')
+    %tbody
+      - orders.each do |order|
+        %tr
+          %td= order.ref
+          %td= order.completed_at
+          %td= number_to_currency(order.total)
+          %td= link_to(t('order.view'), order)
```
Comment (yamnich) [inline app/views/orders/_order_details.html.haml]: move to partial

File: app/views/orders/_order_in_progress.html.haml [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+          %td
+            = image_tag(order_book.book.image.thumb, class: 'img-responsive')
+          %td
+            = link_to(order_book.book_title, book_path(order_book.book))
+          %td
+            = number_to_currency(order_book.price)
+          %td
+            = order_book.quantity
+          %td
+            = number_to_currency(order_book.subtotal)
```
Comment (yamnich) [inline app/views/orders/_order_in_progress.html.haml]: move to partial

File: spec/controllers/books_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    
+    it 'return book' do 
+      expect(assigns(:book)).to eq(book)
+    end
+    
+    it 'render show view' do
+      expect(response).to render_template(:show)
+    end
+  end
+end
```
Comment (yamnich) [inline spec/controllers/books_controller_spec.rb:30]: tests for redirect if user is not authorized

File: spec/features/add_review_spec.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  scenario 'user try add not valid review' do
+    sign_in(user)
+    visit(book_path(book))
+    click_on('Leave a Review')
+    
+    click_on('Add')
+    
+    expect(page).to have_content("Text review can't be blank")
+    
+
```
Comment (yamnich) [inline spec/features/add_review_spec.rb]: remove empty lines

File: app/forms/checkout_form.rb [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+  attr_accessor :order,
+                :step,
+                :use_billing,
+                :shipment_id,
+                :billing_address_attributes,
+                :shipping_address_attributes,
+                :credit_card_attributes
+
+  validate :create_errors
```
Comment (yamnich) [inline app/forms/checkout_form.rb:12]: validate that we have enough items

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (yershovkiril) [inline app/controllers/users_controller.rb]: done, I moved method 'update_password' to devise controller

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (yershovkiril) [inline app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (yershovkiril) [inline app/controllers/users_controller.rb]: we can :) I removed unnecessary method.
P.S. I don't know why, but I thought that we can't use 'current_user' in view.

File: app/forms/checkout_form.rb (hunk 5 above)
Comment (yershovkiril) [inline app/forms/checkout_form.rb]: done

File: app/helpers/application_helper.rb (hunk 6 above)
Comment (yershovkiril) [inline app/helpers/application_helper.rb]: removed

File: app/models/ability.rb (hunk 7 above)
Comment (yershovkiril) [inline app/models/ability.rb:15]: done

File: app/models/author.rb (hunk 8 above)
Comment (yershovkiril) [inline app/models/author.rb]: done

File: app/views/books/show.html.haml (hunk 9 above)
Comment (yershovkiril) [inline app/views/books/show.html.haml]: done, I added i18n all place where I missed there.

File: app/views/devise/mailer/confirmation_instructions.html.erb (hunk 10 above)
Comment (yershovkiril) [inline app/views/devise/mailer/confirmation_instructions.html.erb]: done

File: app/views/orders/_order_details.html.haml (hunk 11 above)
Comment (yershovkiril) [inline app/views/orders/_order_details.html.haml]: done

File: app/views/orders/_order_in_progress.html.haml (hunk 12 above)
Comment (yershovkiril) [inline app/views/orders/_order_in_progress.html.haml]: done

File: spec/controllers/books_controller_spec.rb (hunk 13 above)
Comment (yershovkiril) [inline spec/controllers/books_controller_spec.rb:30]: I added test to review_controller_spec

File: spec/features/add_review_spec.rb (hunk 14 above)
Comment (yershovkiril) [inline spec/features/add_review_spec.rb]: done

File: app/forms/checkout_form.rb (hunk 15 above)
Comment (yershovkiril) [inline app/forms/checkout_form.rb:12]: I added validations when user adds book to cart.

**Pull request merged**